lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,712 @@
+        widget = child,
+        row = row,
+        col = col,
+        row_span = rspan,
+        col_span = cspan
+    }
+    table.insert(self._private.widgets, child_data)
+    matrix_add(self._private.matrix, child_data, row, col, rspan, cspan)
+
+    -- Update the row and column numbers
+    if self._private.num_rows < row + rspan - 1 then
+        self._private.num_rows = row + rspan - 1
+    end
+    if self._private.num_cols < col + cspan - 1 then
+        self._private.num_cols = col + cspan - 1
+    end
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+    return true
+end
+
+--- Add some widgets to the given grid layout
+-- the widgets are assumed to span one cell
+-- @param ... Widgets that should be added (must at least be one)
+function grid:add(...)
+    local args = { n=select('#', ...), ... }
+    assert(args.n > 0, "need at least one widget to add")
+    local row, column
+    for i=1, args.n do
+        local w = args[i]
+        -- Get the next empty coordinate to insert the widget
+        row, column = self:get_next_empty(row, column)
+        if row ~= nil and column ~= nil then
+            self:add_widget_at(w, row, column, 1, 1)
+        else
+            -- TODO: add error message
+            return
+        end
+    end
+end
+
+--- Find the next available cell to insert a widget
+-- @tparam number row The row coordinate of the last occupied cell (default to 1)
+-- @tparam number col The column coordinate of the last occupied cell (default to 1)
+-- @return number,number The row,column coordinate of the next empty cell
+function grid:get_next_empty(hint_row, hint_column)
+    local row    = (hint_row and hint_row > 0) and hint_row or 1
+    local column = (hint_column and hint_column > 0) and hint_column or 1
+    local next = next
+    local matrix = self._private.matrix
+
+    -- Add new widget to next column. Create new row if needed
+    if self._private.orientation == "y" then
+        -- Scan rows until the end of time
+        local current_column = column
+        for i = row, math.huge do
+            if type(matrix[i]) ~= "table" then
+                matrix[i] = {}
+            end
+            if next(matrix[i]) == nil then
+                return i, 1
+            end
+            -- Scan columns until the maximum column number
+            for j = current_column, self._private.num_cols do
+                if matrix[i][j] == nil then
+                    return i, j
+                end
+            end
+            current_column = 1
+        end
+    -- Add new widget to next row. Create new column if needed
+    elseif self._private.orientation == "x" then
+        -- Scan cols until the end of time
+        local current_row = row
+        if type(matrix[current_row]) ~= "table" then
+            matrix[current_row] = {}
+        end
+        if next(matrix[current_row]) == nil then
+            return current_row, column
```

Comment:
I do not really like the code duplication between `x` and `y`. What I actually want to say: The other case has another outer `for` loop and returns `i,1` in this case. Why the (completely?) different behaviour here?